FocusManager: Fix focus_flags check in moveFocusTo, and deal with the fallout

NiLuJe opened the pull request:
* FocusManager: Fix `focus_flags` check in `moveFocusTo` (0 is truthy in Lua, can't do AND checks like in C ;).)
* FileManager+FileChooser: Pass our custom title bar directly to FileChooser (which also means we can now use FC's FocusManager layout directly).
* FileChooser/Menu: Get rid of the weird `outer_title_bar` hack, and simply take a `custom_title_bar` pointer to an actual TitleBar instance instead.
* FileManager/Menu/ListMenu/CoverMenu: Fix content height computations in `_recalculateDimen` (all the non-FM cases were including an old and now unused padding value, `self.header_padding`, leading to more blank space at the bottom than necessary, and, worse, leading to different item heights between FM views, possibly leading to unnecessary thumbnail scaling !)
* ButtonDialog: Proper focus management when the ButtonTable is wrapped in a ScrollableContainer.
* ConfigDialog: Implement a stupid workaround for a weird FocusManager issue when going back from `[⋮]` buttons.
* ConfigDialog: Don't move the visual focus in `update` (i.e., we use `NOT_FOCUS` now that it works as intended).
* DictQuickLookup: Ensures the `Menu` key bind does the exact same thing as the hamburger icon.
* DictQuickLookup: Ensure we refocus after having mangled the FocusManager layout (prevents an old focus highlight from lingering on the wrong button).
* FileChooser: Stop flagging it as no_title, because it is *never* without a title. (This behavior was a remnant of the previous FM-specific title bar hacks, which are no longer a thing).
* FileChooser: Stop calling `mergeTitleBarIntoLayout` twice in `updateItems`. We already call Menu's, which handles it. (Prevents the title bar from being added twice to the FocusManager layout).
* FocusManager: Relax the `Unfocus` checks in `moveFocusTo` to ensure we *always* unfocus something (if unfocusing was requested), even if we have to blast the whole widget tree to do so. This ensures callers that mangle self.layout can expect things to work after calling it regardless of how borked the current focus is.
* FocusManager: Allow passing `focus_flags` to `refocusWidget`, so that it can be forwarded to the internal `moveFocusTo` call.
* FocusManager: The above also allows us to enforce a default that ensures we do *not* send a Focus event on Touch devices, even if they have the hasDPad devcap. This essentially restores the previous/current behavior of not showing the visual feedback from such focus "events" sent programmatically, given the `focus_flags` check fix at the root of this PR ;).
* InputDialog: Fix numerous issues relating to double/ghost instances of both InputText and VirtualKeyboard, ensuring we only ever have a single InputText & VK instance live.
* InputDialog: Make sure every way we have of hiding the VK play nice together, especially when the `toggleKeyboard` button (shown w/ `add_nav_bar`) is at play. And doubly so when we're `fullscreen`, as hiding the VK implies resizing the widget.
* InputText: Make sure we're flagged as in-focus when tapping inside the text field.
* InputText: Make sure we don't attempt to show an already-visible VK in the custom `hasDPad` `onFocus` handler.
* Menu: Get rid of an old and no longer used (nor meaningful) hack in `onFocus` about the initial/programmatically-sent Focus event.
* Menu: Get rid of the unused `header_padding` field mentioned earlier in the FM/FC fixes.
* Menu: Use `FOCUS_ONLY_ON_NT` in the explicit `moveFocusTo` call in `updatePageInfo`, so as to keep the current behavior of not showing the visual feedback of this focus on Touch devices.
* Menu: Make sure *all* the `moveFocusTo` calls are gated behind the `hasDPad` devcap (previously, that was only the case for `updatePageInfo`, but not `mergeTitleBarIntoLayout` (which is called by `updateItems`).
* MultiInputDialog: Actively get rid of the InputText & VK instances from the base class's constructor that we do not use.
* MultiInputDialog: Ensure the FocusManager layout is *slightly* less broken (password fields can still be a bit weird, though).
* TextViewer: Get rid of the unfocus -> layout mangling -> refocus hack now that `refocusWidget` handles this case sanely.
* VirtualKeyboard: Notify our parent InputDialog when we get closed, so it can act accordingly (e.g., resize itself when `fullscreen`).
* ScrollableContainer: Implement the necessary machinery for focus handling inside ButtonDialog (specifically, when scrolling via PgUp/PgDwn).
* TextEditor: Given the above fixes, the plugin is no longer disabled on non-touch devices.
* ReaderBookMark: Make sure we request a full refresh when closing the "Edit note" dialog, as CRe highlights may extend past its dimensions, and if it's closed separately from VK, the refresh would have been limited to its own dimensions, leaving a neat InputDialog-sized hole in the highlights ;).
 
c.f., https://github.com/koreader/koreader/pull/12359#issuecomment-2292651494

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/12361)
<!-- Reviewable:end -->

NiLuJe commented:
As mentioned in #12359, this means that, on `hasDPad` devices, the initial focus is now *always* visible (except in TouchMenu & ConfigDialog, where the tab selection make it obvious enough that we skip it).

![Screenshot_20240816_053820](https://github.com/user-attachments/assets/d126fa05-3845-446a-b17a-9ab735ac65fa)
![Screenshot_20240816_053838](https://github.com/user-attachments/assets/82f711ae-dbef-4ebb-9b59-3f18f16d8db1)

So, err, RFC, I guess?

It certainly looks funky on the emu, but it actually makes groking what the hell is happening with FocusManager much much much clearer, so I suspect it's a net win for actual UX on NT devices?

Frenzie commented:
You can always gate it behind notisTouchDevice.

Commodore64user commented:
So this appear to be working fine as far as I can see, except for a couple of cases (so far). Number 1 is `History` where the `X` in the top right corner will remain inverted (even though is not in focus) until you cycle through it, in which case it goes back to normal. see #9041. this might disappeared as an issue from kindle with #12328 but I am not running that nightly yet, in any case other `hasDPad` that are not kindle would still be affected.

More deceiving though is the dictionary one where the wrong element is highlighted, as you see in the pics below, one would thing `Add to vocab builder` is in focus but if you press <kbd>press</kbd> on it, menu pops up. see #11803 and (I can't find a follow up to this one that fix something or other)

<img src="https://github.com/user-attachments/assets/ea789de9-eced-4ec9-a23c-6add51bcb1fb" width=40% height=40%>
<img src="https://github.com/user-attachments/assets/fc0350b6-4d69-4bcf-8dc5-b5fe0ddb1cb9" width=40% height=40%>

NiLuJe commented:
@Commodore64user: Thanks, fixed both of those ;).

As for SortWidget, that requires #12359, too ;).

Commodore64user commented:
thanks!

that hamburger menu could go the way of the dodo too, <kbd>menu</kbd>  button opens it now so not really necessary on kindle.

```diff
+ if not Device:hasSymKey() or Device:hasScreenKB() then
      table.insert(self.button_table.layout, 1, { self.dict_title.left_button });
+ end

NiLuJe commented:
> that hamburger menu could go the way of the dodo too, menu button opens it now so not really necessary on kindle.

That would require more care because it affects the index at which the next insert needs to be. And the actual menu key on keyboards isn't bound to it anyway, so, eeeh, too much device-specific crap.

Commodore64user commented:
> > that hamburger menu could go the way of the dodo too, menu button opens it now so not really necessary on kindle.
> 
> That would require more care because it affects the index at which the next insert needs to be. And the actual menu key on keyboards isn't bound to it anyway, so, eeeh, too much device-specific crap, pass.

Roger that.

Commodore64user commented:
and if 

> Also, Menu's default linesize is line.medium, which is generally so damn
tiny (1px) the switch from dark gray to black is barely perceptible.

is a thought spoken out loud, I agree, maybe it should be `line.thick` too

NiLuJe commented:
> is a thought spoken out loud, I agree, maybe it should be `line.thick` too (at the very least)

The issue is that, unlike in TouchMenu, in Menu, it serves as the item separator, too; And the current thickness works quite well for that.

[48289d9](https://github.com/koreader/koreader/pull/12361/commits/48289d992a84c9cef6e6580f67918bf17e88f572) attempts to swap it just for the focused item.

Commodore64user commented:
okay, so in classic mode, the unfocus does not clear the first position correctly

<p align="center">
<img src="https://github.com/user-attachments/assets/43ac9bbc-9aa2-4b07-96fa-36b64459d766" width=40% height=40%> 
</p>

and it might be wise to use a similar trick for `detailed list` as well where the focus is also thin af.

Commodore64user commented:
sorry if I am pushing my luck now but, here is another one, twofold really

<p align="center">
<img src="https://github.com/user-attachments/assets/a11463a9-93b5-4bf8-9a91-7d45dccebb24" width=40% height=40%> 
</p>

first the obvious huge gap at the bottom, for reference this does not happen on kindle voyage (when one uses the page turn buttons to scroll down the list) and second the focus does not move along with the scroll, that is to say, if there are 30 results you still need to press the button many many times to even see where the focus is, for reference in that pic focus is still in the very first result in the list.

Commodore64user commented:
> And the actual menu key on keyboards isn't bound to it anyway

Just re-read this and just wanted to say, yes it is.

https://github.com/koreader/koreader/blob/cfd42254acd341609d0927af0cf48c8fcec08fa0/frontend/ui/widget/menu.lua#L966

NiLuJe commented:
> Just re-read this and just wanted to say, yes it is.

Oh, it's because we mind Menu to F1 on SDL, and Menu actually gets called ContextMenu (which gets coopted by FocusManager for... Hold?)

NiLuJe commented:
For clarity, I mean KEY_MENU here, e.g., the thing right next to RCtrl ;).

Frenzie commented:
> which gets coopted by FocusManager for... Hold

Yes, I don't think this should be surprising. ;-) It results in a very intuitive user experience. At the moment we're missing Shift+F10 support for keyboards without KEY_MENU though.

NiLuJe commented:
> for some reason the loop (classic mode) now includes the header items (home and plus menu) twice. just press down a lot and you'll see

That applies to master, too, FWIW.

NiLuJe commented:
> > for some reason the loop (classic mode) now includes the header items (home and plus menu) twice. just press down a lot and you'll see
> 
> That applies to master, too, FWIW.

And what a doozy that turned out to be.. -_-".

We had two different TitleBar in there, because Menu insisted on creating one, but FileManager wanted to display its own, so there was an hilariously broken mix of hacks to display one and not the other (maybe, mostly), but both ended up seepign through to FocusManager...

NiLuJe commented:
> And what a doozy that turned out to be.. -_-".
> 
> We had two different TitleBar in there, because Menu insisted on creating one, but FileManager wanted to display its own, so there was an hilariously broken mix of hacks to display one and not the other (maybe, mostly), but both ended up seepign through to FocusManager...

MVP goes to the default titlebar having a close button, because I noticed that there was a ghost widget in the cycle (conveniently at 1,1), and introspection revealed `icon = "close"` which made me go: WTF?!

Commodore64user commented on frontend/ui/widget/menu.lua:
```diff
@@ -1131,21 +1123,24 @@ function Menu:mergeTitleBarIntoLayout()
-            table.insert(self.layout, menu_item_layout_start_row, row)
-            menu_item_layout_start_row = menu_item_layout_start_row + 1
-        end
+    -- Menu uses the right key to trigger the context menu: we can't use it to move focus in horizontal directions.
```
For reference, this is no longer the case, unless you have hasFewKeys, all other devices have full dPad navigation functionality.

NiLuJe replied:
I took that to mean that Menu (the class) binds Menu (the key), so it cannot uses FocusManager's own Menu bind.

Which is true (on SDL, because FocusManager binds Contextmenu, which we discussed a few hours ago.. somewhere ;p).

NiLuJe replied:
Oh, it's talking about the `Right` key, lol.

I clearly need more sleep.

Commodore64user commented:
> because I noticed that there was a ghost widget in the cycle

this sounds remarkably similar to what happens with multiinputtext. There is a random ghost ether-like step in the cycle as seen here https://github.com/koreader/koreader/issues/9093#issuecomment-2292428320

NiLuJe commented:
> sorry if I am pushing my luck now but, here is another one, twofold really
> 
> <img alt="" width="40%" height="40%" src="https://private-user-images.githubusercontent.com/97603719/358725747-a11463a9-93b5-4bf8-9a91-7d45dccebb24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM5MTg1NDEsIm5iZiI6MTcyMzkxODI0MSwicGF0aCI6Ii85NzYwMzcxOS8zNTg3MjU3NDctYTExNDYzYTktOTNiNS00YmY4LTlhOTEtN2Q0NWRjY2ViYjI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE3VDE4MTA0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2ZjgyMjJmMjIxODRhN2IyMjlmMTE2YmMyNGUzMmJmMGFmYzJmNzdlMDU2M2IzMGYxZThjYzNlMWE4MGJhNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zJap0SjGVCVG2PhJ5-Klti_RdUfUEF5rQsrkxXtAQok">
> 
> first the obvious huge gap at the bottom, for reference this does not happen on kindle voyage (when one uses the page turn buttons to scroll down the list) and second the focus does not move along with the scroll, that is to say, if there are 30 results you still need to press the button many many times to even see where the focus is, for reference in that pic focus is still in the very first result in the list.

That... wasn't pretty ;).

I'm woefully unfamiliar with those widgets, so if @poire-z has any better ideas on how to handle ButtonDialog's ScrollableContainer given that the only thing that's a FocusManager is the ButtonTable, feel free to look at my horrible hack in [d550610](https://github.com/koreader/koreader/pull/12361/commits/d55061049163a43c861707116d4d5797ba085887) ;).

poire-z commented:
> here is another one, twofold really
> 
> <img alt="" width="40%" height="40%" src="https://private-user-images.githubusercontent.com/97603719/358725747-a11463a9-93b5-4bf8-9a91-7d45dccebb24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM5NjQ4NzMsIm5iZiI6MTcyMzk2NDU3MywicGF0aCI6Ii85NzYwMzcxOS8zNTg3MjU3NDctYTExNDYzYTktOTNiNS00YmY4LTlhOTEtN2Q0NWRjY2ViYjI0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE4VDA3MDI1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4NzY1NWMyZDVhMjcyZGE3MGYzMmZlOGY2MTkzNDhhYjc1NzM4ZmM1MTcyZTU5MGMxOTgzMWFlNzgyMTdmNDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.r-GsFal05pZYiYRXqu9TMcoEHpFt6R6PqSawkrkGJAc">
> 
> first the obvious huge gap at the bottom, for reference this does not happen on kindle voyage (when one uses the page turn buttons to scroll down the list)

That huge gap is actually a "feature":
On this screenshot, I think we got 30 results, but the height allows for showing only 18 results "per page". So, we show 18 results on the first page. When the user scrolls down, we show the "remaining" resuults, so only 12, and we start on the 19th result at top, so 6 "holes" at the bottom.
If we were not, and we would anchor the last result at the bottom, we would show 18 results and no blank gap, but the user would have to find out where to start reading results again and skip the 6 he has already glanced at.
We thought about OCD users: they can just scroll down once more, and the whole thing will be anchored at the bottom.
The "solution" would be to guess (might not be possible) the height this ButtonTable will be, and adjust the number of results so it's a multiple of the nb of slots. I have the feeling I added something like that or the other way around, requesting 30 results because then, we can show 5 or 10 or 15 results per page, without blank - and forcing the button table to be exactly the height of 5 or 10 or 15 results. (May be this wasn't ensure on the tall dict lookup?)

> I'm woefully unfamiliar with those widgets, so if @poire-z has any better ideas on how to handle ButtonDialog's ScrollableContainer given that the only thing that's a FocusManager is the ButtonTable, feel free to look at my horrible hack in [d550610](https://github.com/koreader/koreader/pull/12361/commits/d55061049163a43c861707116d4d5797ba085887) ;).

Well, I have become unfamiliar with them :)
(And I'd like to keep my sanity and stop following @Commodore64user esoteric adventures in our obscure parts of the code. I'm glad you can't be focused enough to dig (into) FocusManager :))

Commodore64user commented:
But @poire-z why then the kindle voyage doesn't have those massive gaps as well when using the page turn button to scroll. There, it seems to accurately adjust the height just fine...

poire-z commented:
Dunno, that might be for you to investigate why you end up with 18, while this should ensure we end up with 15, 10 or 6 :/:
https://github.com/koreader/koreader/blob/91c15d52778708b580e0bd0bcc6e364051e98cbc/frontend/ui/widget/dictquicklookup.lua#L1560-L1567
https://github.com/koreader/koreader/blob/91c15d52778708b580e0bd0bcc6e364051e98cbc/frontend/ui/widget/buttondialog.lua#L167-L185

Commodore64user commented:
How fun, you can get it in the opposite end as well. replicate: double <kbd>PgFwd</kbd> and <kbd>PgBack</kbd> 

<p align="center">
<img src="https://github.com/user-attachments/assets/9a6e7f64-08f2-4e77-8fc4-097e0401fd0c" width=40% height=40%> 
</p>

NiLuJe commented:
> Dunno, that might be for you to investigate why you end up with 18, while this should ensure we end up with 15, 10 or 6 :/:

AFAICT, it may or may not have to do with passing `true` as the last argument to `_scrollBy`.

One the other hand, it may also have to do with something else, because I now do that for single-step focus moves *and* the pageturn buttons, and the single-step will re-snap and scroll-up instead of filling the bottom with blanks :D.
(So the only difference in that one request a scroll by the full height of the scrollwidget, vs. an "y offset to next button" value).

In any case, I just realized that this broke navigating other stuff like SpinWidgets (with arrow keys but not tabs, lulz) for some reason, so, back to drawing board ^^.

NiLuJe commented:
> AFAICT, it may or may not have to do with passing `true` as the last argument to `_scrollBy`.
> [...]

Okay, must have last tested *before* switching single steps to `true`.

So, yeah, it's definitely that.

NiLuJe commented:
> TL;DR: Working as intended, it ensures that a jump will only move the single element at the _edge_ off-screen (be that top or bottom), so you don't actually get any overlap in the content while still not jumping over anything, which implies filling the extra space on the edge with blanks.

Note that you still get a snap-back with single-steps, because those will never ask to scroll past the *content* boundaries (which is effectively what *can* happen with the Pg key binds, as they blindly throw a *container's* full of height at it, so if there's less content than a container's worth of height, you get padding ;).

Commodore64user commented:
> what can happen with the Pg key binds, as they blindly throw a container's full of height at it

but why doesn't the voyage do the same thing with its Pg keys?

NiLuJe commented:
> but why doesn't the voyage do the same thing with its Pg keys?

It probably does, it's the exact same codepath. You might not be seeing it just because of content layout changes because of the different resolution?

NiLuJe commented:
> this sounds remarkably similar to what happens with multiinputtext. There is a random ghost ether-like step in the cycle as seen here [#9093 (comment)](https://github.com/koreader/koreader/issues/9093#issuecomment-2292428320)

I have no idea how you're supposed to navigate a MultiInputDialog (in particular: why the hell isn't hiding the VK bound to the actual keyboard key on the K4?!).

Ping @Commodore64user for more context, as I can't test this on SDL (because SDL is a special snowflake in inputText), and I'm pretty sure I broke something related to it on a K4, but I already couldn't figure out how to navigate it on master, so it's only slightly more broken ^^.

NiLuJe commented:
> it might be wise to use a similar trick for `detailed list` as well where the focus is also thin af.

There's a drawback to this trick though: it displaces the text, which leads to a visible refresh of the text. It was only mildly annoying on the emu, but it might be unbearabe no an actual eInk device ;).

Commodore64user commented:
> I have no idea how you're supposed to navigate a MultiInputDialog

if you hide the virtual keyboard `[  ] show virtual keyboard`, is just as simple as pressing up/down. else, you need to unfocus first from inputtext and then use the cursor keys. In any case, it proves the point of being badly broken.

> in particular: why the hell isn't hiding the VK bound to the actual keyboard key on the K4?!

because is a mod key, and... can mod keys have events/shortcuts of their own?

NiLuJe commented:
> because is a mod key, and... can mod keys have events/shortcuts of their own?

Ugh ;'(.

Commodore64user commented:
> > because is a mod key, and... can mod keys have events/shortcuts of their own?
> 
> Ugh ;'(.

in all, is better that way. lots of lovely things would not have been possible if that weren't the case.

NiLuJe commented:
Okay, then, yeah, fix works without the VK shown.

Shit is extra crispy broken without that, though (i.e.:, the defaults):

Back does... wonky things? And Kb + Home does... even wonkier things. (I would expect it to toggle he VK regardless of the "show VK" option?).

NiLuJe commented:
> Shit is extra crispy broken without that, though (i.e.:, the defaults):

Most likely because there are two VKs stacked on each other -_-".

```
08/19/24-01:39:43 DEBUG painting widget: table: 0x4269bd40 
08/19/24-01:39:43 DEBUG painting widget: VirtualKeyboard 
08/19/24-01:39:43 DEBUG painting widget: VirtualKeyboard
```

Commodore64user commented:
> I would expect it to toggle the VK regardless of the "show VK" option?

yes it should always show/hide independently of the setting, which is why it lives alone on `inputtext` and does not report to `vk` or `input dialogue`

> (Testing on the OPDS edit screen, as it's fairly easy to access on stock configs and doesn't require WiFi).

I normally use the KOReader text thing from status bar, but that works too I guess. for extra fun there is one with like 5 fields in `cloud storage > MENU key > WebDAV` also does not require wifi to get there.

NiLuJe commented:
Yeah, the wonkiness all came from the double VK, should be fixed now.

With the exception of password fields being wonkly when there's a password toggle being shown, but I don't even see it on the K4 for some reason, so, meh.

NiLuJe commented:
Marking as ready and flagging for release because this uncovered Lovecraftian monstrosities.

Commodore64user commented:
If the thickness of the focus can’t be increased, what about reducing the tonality/greyness of the separators? If they are less closer to black, the switch might be more noticeable.

Commodore64user commented on frontend/ui/widget/dictquicklookup.lua:
```diff
@@ -726,10 +726,12 @@ function DictQuickLookup:init()
     if Device:hasDPad() then
-        table.insert(self.button_table.layout, 1, { self.dict_title.left_button });
-        table.insert(self.button_table.layout, 2, { lookup_edit_button });
+        table.insert(self.button_table.layout, 1, { self.dict_title.left_button })
```
it appears that selecting this option with focus manager and <kbd>press</kbd>, and doing it by pressing <kbd>Menu</kbd> yields two different results, one is nice and orderly (FM one), the other other one is the same mess from before (Menu key bound event).

Commodore64user replied:
<p align="center">
<img src="https://github.com/user-attachments/assets/1ced504e-3028-4c73-aefc-a12d2c88b356" width=40% height=40%>
<img src="https://github.com/user-attachments/assets/df10dae8-245e-4795-b9a9-77349a8fddd3" width=40% height=40%> 
</p>

Commodore64user replied:
This stuff is still a bit rough and jagged, it acts really funny (not the hilarious kind) when using cursor keys. If you can try it on k4 it would be best cause i can’t even begin to explain how unpredictable it is.

NiLuJe replied:
Is that widget used somewhere that *doesn't* require network connectivity?

Commodore64user replied:
I don’t think you would get as many results anywhere else. Unless you have like 20 dictionaries that is.

Commodore64user replied:
I think this explains the discrepancy between using Focus manager and pressing the menu button

https://github.com/koreader/koreader/blob/ab47bd10a4b65a7ad5161b80fbcbbbf03689c2aa/frontend/ui/widget/dictquicklookup.lua#L232-L237

NiLuJe commented on frontend/ui/widget/multiinputdialog.lua:
```diff
@@ -104,6 +104,12 @@ local MultiInputDialog = InputDialog:extend{
     -- init title and buttons in base class
     InputDialog.init(self)
+    -- Reset layout, we're not using InputDialog's own field
+    self.layout = {}
```
Oh, good point, I forgot that buttons could be a thing.

Commodore64user replied:
> it looks much better in practice on an actual 800x600 eink screen

still kinda rubbish though ;)

NiLuJe replied:
On second thought: what buttons? :D. InputDialog never adds buttons to this anywhere? (Except for the weird case-sensitive thingy from... whatever is actually using that).

Commodore64user replied:
I don’t really know how that worked before as it was clearly a complete mess, but it might be that they were being inherited, now that you clear that at the start they don’t show up anymore, obviously.

NiLuJe replied:
Oh, those buttons, lol.

Whelp, no idea how they got there, but, yeah, fixed it ;p.

Commodore64user replied:
I think the inheritance makes sense since the cycle went (field 1, buttons and then it would finally start with multi input fields)

NiLuJe commented on frontend/ui/widget/menu.lua:
```diff
@@ -475,25 +475,19 @@ function MenuItem:getDotsText(face)
-    end
+function MenuItem:onFocus()
+    self._underline_container.color = Blitbuffer.COLOR_BLACK
+    -- NOTE: Medium is really, really, really thin; so we'd ideally swap to something thicker...
```
Possibly, but that's a potential discussion for another PR ;).

FWIW, I found that it looks much better in practice on an actual 800x600 eink screen than it did on the 800x600 emu ;p.

Commodore64user replied:
> it looks much better in practice on an actual 800x600 eink screen

still kinda rubbish though ;)

Commodore64user commented on frontend/ui/widget/multiinputdialog.lua:
```diff
@@ -136,6 +142,8 @@ function MultiInputDialog:init()
         }
         table.insert(self.input_fields, input_field_tmp)
+        --- @fixme: This is semi-broken when text_type is password, as we actually end up with the checkbox instead of the field,
+        --          and a "Press" on the ehckbox will actually focus the password field and *not* check the box.
```
Typo

Commodore64user commented:
> Dunno, that might be for you to investigate why you end up with 18, while this should ensure we end up with 15, 10 or 6 :/:
> 
> https://github.com/koreader/koreader/blob/91c15d52778708b580e0bd0bcc6e364051e98cbc/frontend/ui/widget/dictquicklookup.lua#L1560-L1567
> 
> 
> https://github.com/koreader/koreader/blob/91c15d52778708b580e0bd0bcc6e364051e98cbc/frontend/ui/widget/buttondialog.lua#L167-L185

This is embarrassing @poire-z, the reason why the wiki thing wasn't enforcing the 15, 10 or 6 rule, was because the Menu button was calling another event `onShowResultsMenu` which apparently does not consider the fact that we could end up with a ton of results, therefore behaves like an uneducated spoilt child. In any case that needs fixing.

but, @NiLuJe, besides the focus not updating when using <kbd>PgFwd</kbd> it appears that `onShowWikiResultsMenu` is actually quite well behaved up to 2024.07-85 (which is the nightly i am running on k4 and none of these changes)

NiLuJe commented:
> This stuff is still a bit rough and jagged, it acts really funny (not the hilarious kind) when using cursor keys. If you can try it on k4 it would be best cause i can’t even begin to explain how unpredictable it is.

Can you elaborate? I can't connect that device, it's a salvaged demo device, it'll get BL'ed.

NiLuJe commented:
On the upside, the `toggleKeyboard` button dragon has been slain, so you get a TextEditor.

(I... have no idea why I disabled it, given that the CSS editor was probably equally borken and yet still enabled ;p).

NiLuJe commented:
> Can you elaborate? I can't connect that device, it's a salvaged demo device, it'll get BL'ed.

Or did you mean the dual column one from the Menu key only? (Because that's gone... well, in the wiki at least ;p).

NiLuJe commented:
> Or did you mean the dual column one from the Menu key only? (Because that's gone... well, in the wiki at least ;p).

Installed some dictionaries on the emu, and, well, that's a standard ButtonDialog, so it works as well as basically every other buttondialogs out there ;p.

Commodore64user commented:
> > Can you elaborate? I can't connect that device, it's a salvaged demo device, it'll get BL'ed.
> 
> Or did you mean the dual column one from the Menu key only? (Because that's gone... well, in the wiki at least ;p).

so, on k4 not running any changes made here. the wiki menu thingy only needs the refocus when one presses <kbd>PgFwd</kbd> and when moving from pos. 15 to 16 it should do the equivalent of <kbd>PgFwd</kbd>, other than that, it is well behaved. nothing funny or fishy looking going on there.

the other one, double column `onShowResultsMenu` is a mess, which is what we see in the screenshots. Only I hadn't realised that it was that event and not the wiki as originally thought. Maybe that double column should also be limited to 15, 10 or 6 results, that might make it act properly, who knows at this point.

Commodore64user commented:
> > This stuff is still a bit rough and jagged, it acts really funny (not the hilarious kind) when using cursor keys. If you can try it on k4 it would be best cause i can’t even begin to explain how unpredictable it is.
> 
> Can you elaborate? I can't connect that device, it's a salvaged demo device, it'll get BL'ed.

searching for `a` on the dictionary might give you enough results if you have multiple dicts, on voyage with three I get 11 entries. with your French ones (and perhaps other ones) you might just get enough for a double pager.

NiLuJe commented:
> searching for `a` on the dictionary might give you enough results if you have multiple dicts, on voyage with three I get 11 entries. with your French ones (and perhaps other ones) you might just get enough for a double pager.

I get 10 entries for `in` (these are all English dictionaries), and that looks fine. This is with the PR applied, mind you, but I don't think it should matter much, this is a plain ButtonDialog, stuff like long-press dialogs on files in the FM should be way more complex than this.

Commodore64user commented:
I don't think 10 is enough so I'll try to describe it.

with the changes now, the widget starts focused on entry 1, if you press up here, it loops back to the bottom but only shows 2 results (pos. 29 and 30) and the rest is empty white space as in the screenshots, if you keep pressing up here, it will slowly fill up the space with new entries. back to the beginning again, if instead of pressing up we press down, on the first press the focus moves to position 2 on the list, but any subsequent press will keep the focus in that second place but the list starts to scroll (I don't think this is desirable)

NiLuJe commented:
Oh, yeah, that's precisely as intended, c.f., https://github.com/koreader/koreader/pull/12361#issuecomment-2295382892

NiLuJe commented:
You can drop the final `true` from the `_scrollBy` calls to see how confusing it can get otherwise ;).

Commodore64user commented:
try the behaviour of last stable on emu prior to all these changes, for the wiki one. that seems better. again it needs the refocus bit fixed and https://github.com/koreader/koreader/pull/12361#issuecomment-2297754274

NiLuJe commented:
> that one does not give you the empty white space.

It technically does, you just can't see it because of the combination of the focus not moving and staying at the top, and the container being perfectly sized for that ;).

NiLuJe commented:
believe me, i know how it looks otherwise, this is how I originally implemented it: it may look "nicer" because it snaps to edges, but in practice it's confusing as hell as you never know where the last entry you focused will end up.

Commodore64user commented:
> but in practice it's confusing as hell as you never know where the last entry you focused will end up.

okay. I believe you

but is there no way to at least not have the empty void when going from pos 1 to ps 20? could we not just jump to the bottom and then have the same behaviour in reverse? i.e stay put in second to last pos. and scroll until we reach the first half

Commodore64user commented:
> > Believe me, I know how it looks otherwise, this is how I originally implemented it: it may look "nicer" because it snaps to edges, but in practice it's confusing as hell as you never know where the last entry you focused will end up.
> 
> What I'm talking about:
> 
> ```diff
> diff --git a/frontend/ui/widget/buttondialog.lua b/frontend/ui/widget/buttondialog.lua
> index 1e236d979..f6d2efb05 100644
> --- a/frontend/ui/widget/buttondialog.lua
> +++ b/frontend/ui/widget/buttondialog.lua
> @@ -311,7 +311,7 @@ function ButtonDialog:onFocusMove(args)
>          local focus = self:getFocusItem()
>          if self.dimen and focus and focus.dimen then
>              local button_y_offset = focus.dimen.y - self.dimen.y - self.top_to_content_offset
> -            self.cropping_widget:_scrollBy(0, button_y_offset, true)
> +            self.cropping_widget:_scrollBy(0, button_y_offset)
>          end
>      end
>  
> diff --git a/frontend/ui/widget/container/scrollablecontainer.lua b/frontend/ui/widget/container/scrollablecontainer.lua
> index 2eec01f27..cf4785747 100644
> --- a/frontend/ui/widget/container/scrollablecontainer.lua
> +++ b/frontend/ui/widget/container/scrollablecontainer.lua
> @@ -644,7 +644,7 @@ function ScrollableContainer:onScrollPageUp()
>      if not self._is_scrollable then
>          return false
>      end
> -    self:_scrollBy(0, -self._crop_h, true)
> +    self:_scrollBy(0, -self._crop_h)
>      self:_notifyParentOfPageScroll()
>      return true
>  end
> @@ -653,7 +653,7 @@ function ScrollableContainer:onScrollPageDown()
>      if not self._is_scrollable then
>          return false
>      end
> -    self:_scrollBy(0, self._crop_h, true)
> +    self:_scrollBy(0, self._crop_h)
>      self:_notifyParentOfPageScroll()
>      return true
>  end
> ```

that is clearly rubbish.

I still think that a fixed list with 15 results where the focus moves all the way down to pos 15 is better, and then if you press down again (to pos 16 or <kbd>PgFwd</kbd> ) pos 16-pos 30 shifts into view. is that clear? which appear to be what 2024.07 is doing. only the focus does not update

Commodore64user commented:
this, 

https://github.com/user-attachments/assets/97a6c8ec-67c2-473f-b400-9ac41a47939b

only with the focus updating properly.

Commodore64user commented:
> That would be significantly more complicated, because we'd have to re-invent half of what ScrollContainer already does for us (specifically, while ScrollContainer is small enough and always on-screen, the backing ButtonDialog is as large as necessary, so the coordinates don't match; and ScrollContainer happens to want the set of coordinates we can fairly easily compute based on the focus itself).

is that the answer to reversing the behaviour https://github.com/koreader/koreader/pull/12361#issuecomment-2297793660 or to the static version seen in the clip?

NiLuJe commented:
> I imagine is to the clip

Yup, the clip ;).

NiLuJe commented:
@hius07: This includes a fix for a subtle technical-debt induced bug in `_recalculateDimen` (for all the Menu classes, ListMenu & MosaicMenu included) that lead to (potentially, usually on high pixel density screens) subtly different item sizes between views with a custom title bar and those without (e.g., FM vs. History).

TL;DR: They were accounting for a `self.header_padding` that's been unused for potentially up to three years ;p.

NiLuJe commented:
Discovered the above because the custom title bar fix made it behave like non-custom views, and I realized that items were slightly smaller when testing on a Clara 2E.

Which, ironically, made item height match everywhere. Except it happened to be *wrong* everywhere ;p.

poire-z commented:
> As mentioned in #12359, this means that, on `hasDPad` devices, the initial focus is now _always_ visible (except in TouchMenu & ConfigDialog, where the active tab visual makes it obvious enough that we skip it).
> So, err, RFC, I guess?
> 
> It certainly looks funky on the emu, but it actually makes groking what the hell is happening with FocusManager much much much clearer, so I suspect it's a net win for actual UX on NT devices?

Yep, not a fan, it makes the emulator (and so, AppImages & co) quite ugly.
It may feel ok in Mosaic/Detailed list/Classic FM/History (even if a bit ugly, but it invites you to use your keyboard, as there, we go for selecting something).
But in other places like TOC/Bookmarks (where we may just go for looking) and seeing the first item underlined feels like it's a separator and that first item some kind of header :/
![image](https://github.com/user-attachments/assets/ebdbf08a-869c-4f32-a61d-ec9fa0a59e9e)

Also ugly with DictLookup, and its hamburger top left button highlighted:
![image](https://github.com/user-attachments/assets/4363d1d1-0aec-42b7-8ecb-beb980f06cfc)

Frenzie commented:
The obvious thing to do is to only show it if it came from a keyboard interaction.

Commodore64user commented on frontend/ui/widget/inputdialog.lua:
```diff
@@ -222,7 +222,7 @@ function InputDialog:init()
         self.deny_keyboard_hiding = true
     end
-    if (Device:hasKeyboard() or Device:hasScreenKB())  and G_reader_settings:isFalse("virtual_keyboard_enabled") then
+    if (Device:hasKeyboard() or Device:hasScreenKB()) and G_reader_settings:isFalse("virtual_keyboard_enabled") then
```
this one has already been corrected.

Commodore64user commented:
> TL;DR: They were accounting for a `self.header_padding` that's been unused for potentially up to three years ;p.

you say it wasn't used:

```
./luajit: plugins/coverbrowser.koplugin/mosaicmenu.lua:886: attempt to perform arithmetic on field 'header_padding' (a nil value)
stack traceback:
	plugins/coverbrowser.koplugin/mosaicmenu.lua:886: in function '_recalculateDimen'
	plugins/coverbrowser.koplugin/main.lua:528: in function 'refreshFileManagerInstance'
	plugins/coverbrowser.koplugin/main.lua:625: in function 'action'
	frontend/ui/uimanager.lua:1016: in function '_checkTasks'
	frontend/ui/uimanager.lua:1462: in function 'handleInput'
	frontend/ui/uimanager.lua:1582: in function 'run'
	./reader.lua:280: in main chunk
	[C]: at 0x0001502c
lipc-wait-event exited normally with status: 0
```

Commodore64user commented:
> Also ugly with DictLookup, and its hamburger top left button highlighted:
> ![image](https://private-user-images.githubusercontent.com/24273478/359407443-4363d1d1-0aec-42b7-8ecb-beb980f06cfc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNTYyMTQsIm5iZiI6MTcyNDE1NTkxNCwicGF0aCI6Ii8yNDI3MzQ3OC8zNTk0MDc0NDMtNDM2M2QxZDEtMGFlYy00MmI3LThlY2ItYmViOTgwZjA2Y2ZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIwVDEyMTE1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1MDJmMjdjZThjZmRkMGM0NmNjNmUxMmE1OGIyOWJiMjk1ZGNkY2Y2ZTQ2OTczZmUzNmY1NmQ0MjllMzE0ZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I-lMnoquculGkgWOml610EImhlGwj2HUugAaYnBvD3E)

It might be best to put the focus on the edit (pencil thing) instead. looks better and is probably more fitting for purpose.

poire-z commented on frontend/ui/widget/inputtext.lua:
```diff
@@ -297,13 +297,19 @@ end
                 do end -- luacheck: ignore 541
             else
-                self:onShowKeyboard()
+                if not self:isKeyboardVisible() then
+                    -- Don't break InputDialog:toggleKeyboard...
+                    local skip_vk = self.parent and self.parent._manual_vk_toggle
+                    if not skip_vk then
+                        self:onShowKeyboard()
+                    end
+                end
```
It feels (not sure) this could be related to a dirty suggestion I gave previously and ended up with having `skip_first_show_keyboard` in InputDialog (dirty, because if the upper widget doesn't explicitely call onShowKeyboard(), and it's supposed to happen on user action, the first user action will then be `skip`ped/ignored.
So, if it is somehow related, feel free to clean my dirty suggestion.

Commodore64user replied:
I maintain that, even with that dragon slain, that button still makes little sense on NT, because, the second you remove focus from a field the VK disappears and as soon as one re-focuses on an inputtext, is back. Just sayin'

furthermore the `elseif` would need to change as well as, the button won't pop up the keyboard (which would need to be an InputDialogue one not an InputText one) otherwise `skip_first_show_keyboard` was in fact pointless and should go.

NiLuJe replied:
> So, if it is somehow related, feel free to clean my dirty suggestion.

Nope, unrelated.

That one's because it's, for some mysterious reason, the caller's responsibility to call inputdialog's onShowKeyboard; instead of just having InputDialog do it on its own via onShow; so basically anything that shows an InputDialog does

```lua
local dialog = InputDialog:new{...}
UIManager:show(dialog)
diualog:onShowKeyboard()
```

Which is mad, granted, but unrelated to our current woes ;p.

Frenzie replied:
I think the idea is probably that some may decide not to show by default, although I'm not really sure why that wouldn't just be a flag.

poire-z commented:
No longer familiar with ScrollContainer/ButtonDialog, but your changes read well, so they must be good :)

The focusmanager & keyboard changes also read well.
(While you're at it, may be you'll want to tackle https://github.com/koreader/koreader/pull/12162#issuecomment-2253572467 ? :))

About the FM/Menu/FileChooser title bar changes, will this change geometry / the remaining height available for the file items ? ie. will this cause Cover thumbnails size change (and cached thumbnails not refreshed - but resized on each display - if their size gets smaller?.

Commodore64user commented:
“And once you’re done here, if you feel like exploring these lands (#11908), be cautious of the many fires and dangers around there.”

poire-z commented:
> > The obvious thing to do is to only show it if it came from a keyboard interaction.
> 
> That's easier said than done given how this mess is handled ;o).

What about keeping a flag (ie. `UIManager.last_input_ev_from_touch`) and acting accordingly ?



> > (While you're at it, may be you'll want to tackle [#12162 (comment)](https://github.com/koreader/koreader/pull/12162#issuecomment-2253572467) ? :))
> 
> Yeah, looking at this again made me even surer than I was before that attempting this would be a massive PITA, no thanks ;).

Really ? Not convinced by my approach ? It worked, thought it would just need more confirmation from a distinct bird view.

NiLuJe commented:
> > The obvious thing to do is to only show it if it came from a keyboard interaction.
> 
> That's easier said than done given how this mess is handled ;o).
> 
> What is humanly feasible is to only _show_ the focus on non-Touch devices, though.

Specifically, I meant for the focus events sent programmatically (i.e., refocusWidget), which comprises most of those "initial" focuses.

That should be functionally equivalent to what you proposed ;).

Frenzie commented:
> What is humanly feasible is to only show the focus on non-Touch devices, though.

Please refer to the very first comment. ;-)

NiLuJe commented:
> Really ? Not convinced by my approach ? It worked, thought it would just need more confirmation from a distinct bird view.

Given my old comments, what I'm concerned about is what happens when there are multiple InputDialogs involved (e.g., MultiInputDialogs), or when weird stuff happens with multiple and/or temporary InputDialogs (possibly TextEdit).

But, yeah, in a perfect world, it'd be as simple as that ;p.

Commodore64user commented:
Sorry to keep throwing ingredients into the pot, but if this catches your fancy #12100 ;) as well...

NiLuJe commented:
> > Really ? Not convinced by my approach ? It worked, thought it would just need more confirmation from a distinct bird view.
> 
> Given my old comments, what I'm concerned about is what happens when there are multiple InputDialogs involved (e.g., MultiInputDialog), or when weird stuff happens with multiple and/or temporary InputDialogs (possibly TextEdit).
> 
> But, yeah, in a perfect world, it'd be as simple as that ;p.

Yeah, tried it, c.f., [2ccf760](https://github.com/koreader/koreader/pull/12361/commits/2ccf7601fe1cbd9794aea0be754ea4166b9767d7).

Once you lose focus, InputText refuses to process text because it doesn't know where to send it when there are multiple instances (because of event propagation rules, they'll get the event in the order they appear in the array).

And if you drop the focus restriction, then either nobody, only one, or everybody gets the event, and I can't really see a sane way out of that.

(I mean, I tried ref-counting InputText to at least bypass the "one instance only" focus limit, but that doesn't help for the other cases).

NiLuJe commented:
Besides said experiment, the last few commits made the toggleKeyboard fix more elegant (and fixed another instance of weird-ass ghost widgets), and implemented the "don't show initial focus highlight on touch devices" thingy.

poire-z commented:
> And if you drop the focus restriction, then either nobody (if the first one stops the event as what currently happens), only one, or everybody gets the event, and I can't really see a sane way out of that.

Sounded good (?): no matter the order, if a inputtext knows its focus start:
- if I'm focused, I handle it, return true
- if I'm not focused, return false

> And, more importantly, you're still left with the fact that once you hit Back, none of them are focused anymore anyway ;).

Which is good ? None should get the keys - but arrow keys would still work (as they do no) so one can navigate and refocus ?
We may not even need ref count: one focused enables SDLTextInput - one unfocused disables it, next focused one re-enables it?

NiLuJe commented:
> second if you then re-focus by clicking, you... still don't get key events handled for some reason? :D.

That's... actually a bug in master. Joy.

NiLuJe commented:
> so one can navigate and refocus ?

At which point the VK pops up automatically ;).

NiLuJe commented:
> * if I'm focused, I handle it, return true

There's *never* a focus once we hide the VK, though.

NiLuJe commented:
Note that I'm taliking about the defaults (i.e., with the VK enabled). I have no idea how the "start with VK hidden" thing interacts with these statements.

But, assuming it only applies for the initial show, I assume it mostly doesn't change anything?

NiLuJe commented:
> > second if you then re-focus by clicking, you... still don't get key events handled for some reason? :D.
> 
> That's... actually a bug in master. Joy.

Fixed.

NiLuJe commented:
> Sounded good

I'm getting a little cross-eyed, so I may be missing the point.

Feel free to poke holes at it, and I'll take a fresh look tomorrow ;).

NiLuJe reviewed (commented):
Reviewed, and changelog written in the top post. Ready to squash ;)!

> Reviewed 1 of 4 files at r2, 1 of 3 files at r4, 1 of 5 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, 2 of 5 files at r17, 3 of 4 files at r21, 5 of 6 files at r22, 1 of 1 files at r23, 3 of 3 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12361#-:-O54fCcDASq1AiIpqDWd:bp7akh)* status: all files reviewed, 8 unresolved discussions (waiting on @Commodore64user and @poire-z)



<!-- Sent from Reviewable.io -->

NiLuJe commented:
Is CircleCI having a boo-boo, or is it unhappy because of the branch being a tiny bit stale?

Frenzie commented:
Depends how stale but a rebase might not be a bad idea regardless?

poire-z reviewed (approved):
Approved from my look around some days ago.
(In the first post (and probable commit messsage), you mention Menu `header_padding` twice.)

NiLuJe commented:
> (In the first post (and probable commit messsage), you mention Menu header_padding twice.)

Yeah, the first time is in context if the FM/FC stuff (i.e., why it made no sense).

And the second is just because M comes after F, and the field actually originates in Menu (which is the parent class of FileChooser) ;p

NiLuJe reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12361)* status: all files reviewed, 7 unresolved discussions (waiting on @Commodore64user and @poire-z)



<!-- Sent from Reviewable.io -->

Commodore64user reviewed (approved):
Besides my personal dislike for the scrolling widget (but that is a different matter altogether), this seems to work as intended. Nice work @NiLuJe

NiLuJe reviewed (commented):
> Reviewed 1 of 1 files at r26, 2 of 2 files at r27, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/12361)* status: all files reviewed, 7 unresolved discussions (waiting on @Commodore64user and @poire-z)



<!-- Sent from Reviewable.io -->
